platform/linux-generic/odp_packet.c

```diff
@@ -414,97 +490,217 @@ static inline odp_packet_hdr_t *add_segments(odp_packet_hdr_t *pkt_hdr,
 		new_hdr->buf_hdr.seg[0].len   = seg_len;
 
 		packet_seg_copy_md(new_hdr, pkt_hdr);
-		new_hdr->frame_len    = pkt_hdr->frame_len + len;
-		new_hdr->unshared_len = pkt_hdr->unshared_len + len;
-		new_hdr->headroom     = pool->headroom + offset;
-		new_hdr->tailroom     = pkt_hdr->tailroom;
+		new_hdr->frame_len = pkt_hdr->frame_len + len;
+		new_hdr->headroom  = pool->headroom + offset;
+		new_hdr->tailroom  = pkt_hdr->tailroom;
+		new_hdr->shared_len = pkt_hdr->shared_len;
 
 		pkt_hdr = new_hdr;
 	} else {
-		int last;
+		seg_entry_t *last_seg;
 
 		/* add into the tail */
 		add_all_segs(pkt_hdr, new_hdr);
 
 		/* adjust last segment length */
-		last = packet_last_seg(pkt_hdr);
-		pkt_hdr->buf_hdr.seg[last].len = seg_len;
+		last_seg      = seg_entry_last(pkt_hdr);
+		last_seg->len = seg_len;
 
-		pkt_hdr->frame_len    += len;
-		pkt_hdr->unshared_len += len;
-		pkt_hdr->tailroom      = pool->tailroom + offset;
+		pkt_hdr->frame_len += len;
+		pkt_hdr->tailroom   = pool->tailroom + offset;
 	}
 
 	return pkt_hdr;
 }
 
-static inline void free_bufs(odp_packet_hdr_t *pkt_hdr, int first, int num)
+static inline int seg_is_link(void *hdr)
 {
-	int i, nfree;
-	odp_buffer_hdr_t *buf_hdr[num];
+	odp_packet_hdr_t *pkt_hdr = hdr;
+
+	return pkt_hdr != pkt_hdr->buf_hdr.seg[0].hdr;
+}
+
+static inline void packet_free_multi(odp_buffer_hdr_t *hdr[], int num)
+{
+	int i;
+	uint32_t ref_cnt;
+	int num_ref = 0;
+
+	for (i = 0; i < num; i++) {
+		/* Zero when reference API has not been used */
+		ref_cnt = odp_atomic_load_u32(&hdr[i]->ref_cnt);
```

Comment:
The previous code wrapped the `ref_cnt` manipulation routines  in `static inline` functions to give more flexibility in the implementation. Not a major point, but I think it would make sense to continue that in this version. For example, in cases when references are not being used it's not really necessary to use an `odp_atomic_load_u32()` to inspect the value of `ref_cnt`.  If you read this as zero, you know there's no reference. It's only when `ref_cnt` is non-zero do you need the atomics.

Reply:
Atomic variables must be accessed through atomics API, also when just loading. Atomic_load() is in most cases just a normal load, but still we should use the proper function...

Wrapping load would not help much (load is a load). Increment and store have already been wrapped into buffer_ref_inc(). I'll wrap fetch_dec (into buffer_ref_dec()) and ref_cnt > 1 (into ref_cnt_not_last().